### PR TITLE
HW-1: Building Blocks of OOP. Part-1

### PR DESCRIPTION
**Pull request by StashiKu**
# 📚 Home Task

The home task should be done using TypeScript.

## Specific Steps

1. Create a `Point` class, which creates 2 dimensional point with coordinates. It
   should contain:
  [x] two instance variables `x` and `y`;
  [x] default constructor which creates a point at the location of (0, 0);
  [x] overloaded constructor (use multiple constructors declaration for Typescript)
    which creates a point by `x` and `y` coordinates;
  [x] `toString()` method should return a `Point` class stringified representation in
    format: `"(x, y)"`;
  [x] `distance()` method should be overloaded (use multiple methods declaration for
    Typescript) with next implementations:
    [x] no args: distance from this point to (0, 0);
    [x] `distance(other: Point)` - distance from this point to a given instance of
      `Point`;
    [x] `distance(x, y)` - distance from this point to a given point (x, y).

2. Create abstract superclass called `Shape`, which contains:
  [x] two protected instance variables: `color` (string), `filled`
    (boolean) and `points` (Point[]);
  [x] overloaded constructor (use multiple constructors declaration for Typescript): a
    constructor that takes a list of `points` and initializes the color to `"green"`
    and filled to `true` by default, and a constructor that takes a list of `points`,
    the `color` and `filled` properties;
  [x] Make sure that the `Shape` has at least 3 points (2 points is just a line).
  [x] `toString()` method that returns `"A Shape with color of xxx and filled/Not
    filled. Points: (x1, y1), (x2, y2)..."`;
  [x] `getPerimeter()` that calculates the perimeter using `Point.distance` method;

3. Create class `Triangle` that takes 3 points as it's vertices. Triangle must inherit
   `Shape` abstract class. `Triangle` should contain:

  [x] a constructor (use multiple constructors declaration for Typescript) which creates
    `Triangle` class using three instances of `Point` class, may also provide color and
    filled properties;
  [x] override `toString()` method, it should return a `Triangle` class stringified
    representation in format `"Triangle[v1=(x1, y1),v2=(x2, y2),v3=(x3, y3)]"`;
  [x] override `getType()` method, which prints `"equilateral triangle"` if all the three
    sides are equal, `"isosceles triangle"` if any two of the three sides are equal, or
    `"scalene triangle"` if all sides are different.

## Evaluation criteria

2. Only some classes were implemented.
3. Some classes were not implemented.
4. Some required methods are missing. All `@ts-nocheck` comments are removed and there are no TypeScript warnings exist after that.
5. All tasks are implemented to a full extent and all tests are passed successfully (`npm run test`). All `@ts-nocheck` comments are removed and there are no TypeScript warnings exist after that.

### CODE REVIEW

File: HW-1-Building-Blocks-of-OOP-Part-1/src/Point.ts [hunk 1]
```diff
@@ -0,0 +1,81 @@
+export class Point {
+    public x: number;
```
Comment (metam0rphus) [inline HW-1-Building-Blocks-of-OOP-Part-1/src/Point.ts:2]: better to have it private

File: HW-1-Building-Blocks-of-OOP-Part-1/src/Point.ts [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+        const b = yY - this.y;
+        let c = Math.sqrt((a ** exponent) + (b ** exponent));
+
+        return this.roundNumber(c);
+    }
+
+    public distance(): number;
+    public distance(point: Point): number;
+    public distance(xX: number, yY: number): number;
+    public distance(...arg: any): number | null {
```
Comment (metam0rphus) [inline HW-1-Building-Blocks-of-OOP-Part-1/src/Point.ts:62]: usage of "any" is bad practice

File: HW-1-Building-Blocks-of-OOP-Part-1/src/Point.ts [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+          if (distance < 0) {
+            distance *= -1
+          }
+
+          return this.roundNumber(distance);
+        }
+
+        // Pythagorean theorem calculations
+        const exponent = 2;
```
Comment (metam0rphus) [inline HW-1-Building-Blocks-of-OOP-Part-1/src/Point.ts:51]: Math.hypot

File: HW-1-Building-Blocks-of-OOP-Part-1/src/Point.ts [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+        if (arg.length === 1 && typeof arg[0] === 'number') {
+            return this.calculate(arg[0], 0);
+        }
+
+        if (arg.length === 2 && typeof arg[0] === 'number' && typeof arg[1] === 'number') {
+            return this.calculate(arg[0], arg[1]);
+        }
+
+        return null;
```
Comment (metam0rphus) [inline HW-1-Building-Blocks-of-OOP-Part-1/src/Point.ts:79]: it shouldn't return null

File: HW-1-Building-Blocks-of-OOP-Part-1/src/Triangle.ts [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+import { Shape } from "./Shape";
+
+export class Triangle extends Shape {
+    private p1: Point
+    private p2: Point
+    private p3: Point
+  
+    constructor(p1: Point, p2: Point, p3: Point);
+    constructor(p1: Point, p2: Point, p3: Point, color: string, filled: boolean)
+    constructor(p1: Point, p2: Point, p3: Point, color?: string, filled?: boolean) {
```
Comment (metam0rphus) [inline HW-1-Building-Blocks-of-OOP-Part-1/src/Triangle.ts:11]: just {super([v1, v2, v3], color, filled);}

File: HW-1-Building-Blocks-of-OOP-Part-1/src/Triangle.ts [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+        super([p1,p2,p3], color = 'green', filled = false );
+
+        this.points = [p1,p2,p3];
+        this.color = color ?? 'green';
+        this.filled = filled ?? true;
+    }
+  
+    public toString(): string {
```
Comment (metam0rphus) [inline HW-1-Building-Blocks-of-OOP-Part-1/src/Triangle.ts:19]: toString(): string {
        const [v1, v2, v3] = this.points;
        return `Triangle[v1=${v1},v2=${v2},v3=${v3}]`
    }

**Review by metam0rphus — CHANGES_REQUESTED** [review]

File: HW-1-Building-Blocks-of-OOP-Part-1/src/Point.ts (hunk 1 above)
Comment (StashiKu) [reply HW-1-Building-Blocks-of-OOP-Part-1/src/Point.ts]: fix

File: HW-1-Building-Blocks-of-OOP-Part-1/src/Point.ts (hunk 2 above)
Comment (StashiKu) [reply HW-1-Building-Blocks-of-OOP-Part-1/src/Point.ts]: fixed

File: HW-1-Building-Blocks-of-OOP-Part-1/src/Point.ts (hunk 3 above)
Comment (StashiKu) [reply HW-1-Building-Blocks-of-OOP-Part-1/src/Point.ts:51]: fixed

File: HW-1-Building-Blocks-of-OOP-Part-1/src/Point.ts (hunk 4 above)
Comment (StashiKu) [reply HW-1-Building-Blocks-of-OOP-Part-1/src/Point.ts]: fixed

File: HW-1-Building-Blocks-of-OOP-Part-1/src/Triangle.ts (hunk 5 above)
Comment (StashiKu) [reply HW-1-Building-Blocks-of-OOP-Part-1/src/Triangle.ts:11]: fixed

File: HW-1-Building-Blocks-of-OOP-Part-1/src/Triangle.ts (hunk 6 above)
Comment (StashiKu) [reply HW-1-Building-Blocks-of-OOP-Part-1/src/Triangle.ts:19]: fixed

**Review by metam0rphus — APPROVED** [review]

**Pull request merged**